### PR TITLE
Filtering sources based on assembly type

### PR DESCRIPTION
**Pull request by abhishkk**
Part 3 of [RFC](https://github.com/Microsoft/vstest-docs/pull/124):
Contains changes for filtering sources based on assembly type.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs [hunk 1]
```diff
@@ -237,14 +238,24 @@ private void SetAdapterLoggingSettings(IMessageLogger messageLogger, IRunSetting
...
             var sourcesForWhichNoDiscovererIsAvailable = new List<string>(sources);
 
             foreach (var discoverer in allDiscoverers)
             {
+                var sourcesToCheck = sources;
+
+                if (discoverer.Metadata.AssemblyType == AssemblyType.Native ||
+                    discoverer.Metadata.AssemblyType == AssemblyType.Managed)
+                {
+                    assemblyTypeToSoucesMap = assemblyTypeToSoucesMap ?? GetAssemblyTypeToSoucesMap(sources);
```
Comment (abhishkk) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs]: spelling

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Discovery/DiscovererEnumeratorTests.cs [hunk 2]
```diff
@@ -31,6 +34,7 @@ public class DiscovererEnumeratorTests
         private DiscoveryResultCache discoveryResultCache;
         private Mock<IRequestData> mockRequestData;
         private Mock<IMetricsCollection> mockMetricsCollection;
+        private Mock<IAssemblyProperties> mockPEReaderHelper;
```
Comment (abhishkk) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Discovery/DiscovererEnumeratorTests.cs]: renaming mockPEReaderHelper to mockAssemblyProperties

File: src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs [hunk 3]
```diff
@@ -51,8 +53,22 @@ public DiscovererEnumerator(IRequestData requestData, DiscoveryResultCache disco
             this.discoveryResultCache = discoveryResultCache;
             this.testPlatformEventSource = testPlatformEventSource;
             this.requestData = requestData;
+            this.assemblyProperties = new AssemblyProperties();
         }
 
+        // Added this to make class testable, needed a PEHeader mocked Instance
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs]: Nit: `needed a AssemblyProperties` or remove comment.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs [hunk 4]
```diff
@@ -237,14 +254,24 @@ private void SetAdapterLoggingSettings(IMessageLogger messageLogger, IRunSetting
...
 
+            IDictionary<AssemblyType, IEnumerable<string>> assemblyTypeToSoucesMap = null;
             var result = new Dictionary<LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities>, IEnumerable<string>>();
             var sourcesForWhichNoDiscovererIsAvailable = new List<string>(sources);
 
             foreach (var discoverer in allDiscoverers)
             {
+                var sourcesToCheck = sources;
+
+                if (discoverer.Metadata.AssemblyType == AssemblyType.Native ||
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs]: Is there a chance discoverer can support both managed and native?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs [hunk 5]
```diff
@@ -277,6 +304,49 @@ private void SetAdapterLoggingSettings(IMessageLogger messageLogger, IRunSetting
...
+
+            if (sources != null && sources.Any())
+            {
+                foreach (string source in sources)
+                {
+                    var sourcesList = IsAssembly(source) ?
+                        assemblyTypeToSoucesMap[assemblyProperties.GetAssemblyType(source)] :
+                        assemblyTypeToSoucesMap[AssemblyType.None];
+
+                    ((List<string>)sourcesList).Add(source);
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs]: Can we use `IList` here?

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Discovery/DiscovererEnumeratorTests.cs [hunk 6]
```diff
@@ -81,7 +86,128 @@ public void LoadTestsShouldNotCallIntoDiscoverersIfNoneMatchesSources()
...
+
+                Assert.IsTrue(NativeDllTestDiscoverer.IsNativeDiscoverTestCalled);
+                CollectionAssert.AreEqual(sources, NativeDllTestDiscoverer.Sources.ToList());
+
+                Assert.IsFalse(ManagedDllTestDiscoverer.IsManagedDiscoverTestCalled);
+                Assert.IsFalse(JsonTestDiscoverer.IsDiscoverTestCalled);
+            }
+            finally
+            {
+                this.ResetDiscoverers();
```
Comment (smadala) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Discovery/DiscovererEnumeratorTests.cs]: Add this to cleanup.

**Review by smadala — APPROVED** [review]

File: src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs (hunk 5 above)
Comment (mayankbansal018) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs]: IList doesn't take a string, & does not have Add method, would need to create  another IList for source.

Comment (mayankbansal018) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs]: done

**Pull request merged**